### ISSUE TITLE
Dependent requests can deadlock on puma if single-threaded

### ISSUE THREAD

**Issue by codehotter** [puma/puma#612]
I absolutely love puma and I'm using it for all my applications. Usually, those applications are threadsafe. Sometimes they are not. When an application is running single threaded, either with threads 1, 1 in puma config or allow_concurrency=false in application config, certain kinds of requests can deadlock.

For example, let's say I have a application that generates a PDF from a page server side. The first request is for the PDF, then the PDF creator tool (e.g. phantomjs) internally does a second request to the app. If you're unlucky, and the second request is routed to the same puma worker, the second request will not be handled. It will either block in rack::lock or in puma itself until the first request is finished. However the first request will not finish until the second request is handled.
Another example is a git http server like grack, which will deadlock if you have pre-receive hooks that make calls to your application.

I realize puma may not be the best application server for a single threaded application but I really like puma and I hope I can standardize our application server on it. With the below patch, and config threads 1, 1, the problem is solved for us because puma workers that are 'busy' don't accept new requests. Therefore the request will be routed to another worker.

However I don't understand puma code that well so I'm worried I've broken some other things with this patch. What about performance for example? (Our installations are rather small.)

```
diff -ur puma-2.10.2.orig/lib/puma/server.rb puma-2.10.2/lib/puma/server.rb
--- puma-2.10.2.orig/lib/puma/server.rb 2014-12-01 08:48:32.322000000 +0100
+++ puma-2.10.2/lib/puma/server.rb      2014-12-01 08:49:48.792000000 +0100
@@ -184,7 +184,7 @@
                 break if handle_check
               else
                 begin
-                  if io = sock.accept_nonblock
+                  if !pool.busy and io = sock.accept_nonblock
                     client = Client.new io, nil
                     pool << client
                   end
@@ -293,7 +293,7 @@
                 break if handle_check
               else
                 begin
-                  if io = sock.accept_nonblock
+                  if !pool.busy and io = sock.accept_nonblock
                     client = Client.new io, @binder.env(sock)
                     pool << client
                   end
diff -ur puma-2.10.2.orig/lib/puma/thread_pool.rb puma-2.10.2/lib/puma/thread_pool.rb
--- puma-2.10.2.orig/lib/puma/thread_pool.rb    2014-12-01 08:48:32.333000000 +0100
+++ puma-2.10.2/lib/puma/thread_pool.rb 2014-12-01 09:45:59.449000000 +0100
@@ -114,6 +114,16 @@
 
     private :spawn_thread
 
+    # A thread may become available after busy leaves the critical section
+    # this means that a request may not be handled even though it can be.
+    # However, since there is only one thread that adds work, the calling thread,
+    # the reverse should never happen.
+    def busy
+      @mutex.synchronize do
+        (@todo.size - @waiting) >= (@max - @spawned)
+      end
+    end
+
     # Add +work+ to the todo list for a Thread to pickup and process.
     def <<(work)
       @mutex.synchronize do
```

**Comment by codehotter** [comment]
Evan, thank you for your comments.

In order to both queue requests and balance them across workers, I think you would need some kind of inter process communication mechanism to hand over clients to another process. On Linux, this might be possible with fd passing over unix domain socket and shared memory.
However a very common setup is to have puma behind a reverse proxy like nginx, which would be able to handle slow clients and queue the requests before they reach puma. In that setup, there's not much harm in disabling slow client, HTTP keep-alive and request queueing in puma.
Could this be a configuration option? If the user wants to balance requests across workers, say if they have a single threaded application, then they know that they need a reverse proxy in front of puma and they can enable the configuration option.
I made an attempt to implement a configuration option here: https://github.com/puma/puma/pull/640 . It's too much to put in line, so I made a pull request. That doesn't mean I think it is ready to be merged. For one, I'm not very confident that I did the right thing, for another, I have no idea what the impact of these changes is on performance, and finally, I didn't implement anything for the case of JRuby and SSL socket because I didn't fully understand what was going on there.
I'm running two non-threadsafe apps with this configuration option now, and while it seems to work, I would like your advice for a robust solution.
